distributed/worker.py

```diff
@@ -3388,11 +3333,8 @@ def done_event():
             for d in self.in_flight_workers.pop(worker):
                 ts = self.tasks[d]
                 ts.done = True
-                if d in cancelled_keys:
-                    if ts.state == "cancelled":
-                        recommendations[ts] = "released"
-                    else:
-                        recommendations[ts] = "fetch"
+                if ts.state == "cancelled":
```

Comment:
This tests ts.state a lot later than before. There's a new test in this PR to verify this is works for tasks that transition during the comms.

Reply:
Was the prior behavior just an optimization to avoid fetching keys that were cancelled in the handle_instructions to gather_dep interstice? Before, we avoided fetching them; now we don't? Seems like a nice thing to add back eventually, but the simplification here is nice.

I'm curious how much a blocked event loop https://github.com/dask/distributed/issues/6325 would make this scenario more likely.

Reply:
The prior behaviour was introduced by #5426 as a response to a deadlock.

Before #5426 you had two use cases:
a1. cancelled during comms
a2. cancelled task is received and implicitly transitioned to released

b1. cancelled in the interstice
b2. cancelled task is not fetched
b3. deadlock

After #5426:
a1. cancelled during comms
a2. cancelled task is received and implicitly transitioned to released

b1. cancelled in the interstice
b2. cancelled task is explicitly transitioned to released

After this PR:
1. cancelled whenever
2. cancelled task is explicitly transitioned to released

Reply:
I really don't think we should care about performance optimizations in this case. Transitions from flight to cancelled should not be that frequent to begin with?

Reply:
I agree that we should remove this optimization if possible. It's not worth it and it didn't feel great to introduce it in the first place.
By now, I trust tests around these edge cases enough that if all is green after removal, we're good to go